### PR TITLE
[#feature] GPPI Validator

### PR DESCRIPTION
**Pull request by vlad-tokarev**
Tool to test GPPI packed model

### CODE REVIEW

File: scripts/Jenkinsfile [hunk 1]
```diff
@@ -137,6 +137,14 @@ pipeline {
                         }
                     }
                 }
+                stage('Build odahuflowctl Docker image') {
+                    steps {
+                        script {
+                            cicdLibrary.buildLegionImage('odahuflow-cli', '.', 'containers/odahuflow-cli/Dockerfile')
+                            cicdLibrary.uploadDockerImage('odahuflow-cli', env.param_stable_release.toBoolean() && env.param_dockerhub_publishing_enabled.toBoolean())
```
Comment (mcdoker18) [inline scripts/Jenkinsfile]: We upload the artifacts on the next 'Upload artifacts' stage.

File: containers/odahuflow-cli/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+FROM ubuntu:18.04
+
+ARG MLFLOW_VERSION=1.3.0
```
Comment (mcdoker18) [inline containers/odahuflow-cli/Dockerfile]: Do we need this constant?

File: containers/odahuflow-cli/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    echo "conda activate base" >> ~/.bashrc
+
+ARG ODAHUFLOW_DIR=/opt/odahuflow
+RUN pip install --upgrade pip
+
+COPY odahuFlow/ ${ODAHUFLOW_DIR}/
+
+WORKDIR ${ODAHUFLOW_DIR}/sdk
+RUN python setup.py sdist
+RUN pip install dist/*.tar.gz
```
Comment (mcdoker18) [inline containers/odahuflow-cli/Dockerfile]: `pip install -e .`?

File: containers/odahuflow-cli/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    /opt/conda/bin/conda clean -tipsy && \
+    ln -s /opt/conda/etc/profile.d/conda.sh /etc/profile.d/conda.sh && \
+    echo ". /opt/conda/etc/profile.d/conda.sh" >> ~/.bashrc && \
+    echo "conda activate base" >> ~/.bashrc
+
+ARG ODAHUFLOW_DIR=/opt/odahuflow
+RUN pip install --upgrade pip
+
+COPY odahuFlow/ ${ODAHUFLOW_DIR}/
+
```
Comment (mcdoker18) [inline containers/odahuflow-cli/Dockerfile]: Split the packages installation into two parts:
First, it is a dependency installation. And the second step is a package installation itself.
It allows using the docker cache properly.

File: odahuFlow/cli/odahuflow/cli/parsers/gppi.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@gppi.command(help='Test GPPI model by running it')
+@click.argument('path', type=click.Path(exists=True, file_okay=False))
+@click.option('--use-current-env/--not-use-current-env', 'use_current_env',
+              help='Use current environment', default=False)
+@click.option('--env-name', '-e', 'env_name', help='Environment name to run GPPI model', type=click.STRING)
+@click.option('--skip-deps/--not-skip-deps', 'skip_deps', help='Not install dependencies', default=False)
+def test(path: str, use_current_env: bool = False, env_name: str = '', skip_deps: bool = False):
+    model = GPPITrainedModelBinary(path, use_current_env, env_name, skip_deps)
```
Comment (mcdoker18) [inline odahuFlow/cli/odahuflow/cli/parsers/gppi.py]: Please, provide a little user documentation.

File: odahuFlow/operator/docs/docs.go [hunk 6]
```diff
@@ -1,5 +1,6 @@
 // GENERATED BY THE COMMAND ABOVE; DO NOT EDIT
 // This file was generated by swaggo/swag at
```
Comment (mcdoker18) [inline odahuFlow/operator/docs/docs.go]: Your sed doesn't work as expected. Please, revert changes in the docs.go, swagger.json and swagger.yaml files.

File: odahuFlow/operator/pkg/odahuflow/k8s.go [hunk 7]
```diff
@@ -33,6 +33,7 @@ const (
 	PackagerResultStep        = "result"
 	TrainerSetupStep          = "setup"
 	TrainerTrainStep          = "trainer"
+	TrainerValidationStep     = "validation"
```
Comment (mcdoker18) [inline odahuFlow/operator/pkg/odahuflow/k8s.go]: Please, add the step into this array https://github.com/odahu/odahu-flow/blob/develop/odahuFlow/operator/pkg/repository/packaging/kubernetes/model_packaging.go#L47

File: scripts/Jenkinsfile [hunk 8]
```diff
@@ -137,6 +137,14 @@ pipeline {
                         }
                     }
                 }
```
Comment (mcdoker18) [inline scripts/Jenkinsfile]: Please, add a e2e robot test, when model validation is failed.
Useful links:
* [Example of test trainer](https://github.com/odahu/odahu-flow/blob/develop/odahuFlow/robot/odahuflow/robot/training_data_helper.py#L61)
* [Usage of the test trainer](https://github.com/odahu/odahu-flow/blob/develop/odahuFlow/tests/e2e/robot/tests/training/traininig_data.robot)

File: containers/odahuflow-cli/Dockerfile (hunk 3 above)
Comment (vlad-tokarev) [reply containers/odahuflow-cli/Dockerfile]: Conda have some problems with editable packages
https://github.com/conda/conda/issues/5861

**Pull request merged**